libiocage/lib/Logger.py

```diff
@@ -80,70 +113,95 @@ def log(self, *args, **kwargs):
         if "level" not in kwargs:
             kwargs["level"] = "info"
 
-        self._print(**kwargs)
-        # self._write(**kwargs)
+        log_entry = LogEntry(logger=self, **kwargs)
+
+        if self._should_print_log_entry(log_entry):
+            self._print_log_entry(log_entry)
+            self.PRINT_HISTORY.append(log_entry)
+        
+        return log_entry
+
+    def verbose(self, message, indent=0, **kwargs):
+        return self.log(message, level="verbose", indent=indent, **kwargs)
+
+    def error(self, message, indent=0, **kwargs):
+        return self.log(message, level="error", indent=indent, **kwargs)
 
-    def verbose(self, message, jail=None, indent=0):
-        self.log(
-            message=message,
-            level="verbose",
-            jail=jail,
-            indent=indent
-        )
+    def warn(self, message, indent=0, **kwargs):
+        return self.log(message, level="warn", indent=indent, **kwargs)
 
-    def error(self,
-              message,
-              jail=None,
-              indent=0):
+    def debug(self, message, indent=0, **kwargs):
+        return self.log(message, level="debug", indent=indent, **kwargs)
 
-        self.log(message, level="error", jail=jail, indent=indent)
+    def spam(self, message, indent=0, **kwargs):
+        return self.log(message, level="spam", indent=indent, **kwargs)
 
-    def warn(self,
-             message,
-             jail=None,
-             indent=0):
+    def screen(self, message, indent=0, **kwargs):
+        """
+        Screen does never get printed to log files
+        """
+        return self.log(message, level="screen", indent=indent, **kwargs)
 
-        self.log(message, level="warn", jail=jail, indent=indent)
+    def redraw(self, log_entry):
 
-    def debug(self,
-              message,
-              jail=None,
-              indent=0):
+        if log_entry not in self.PRINT_HISTORY:
+            raise libiocage.lib.errors.CannotRedrawLine(
+                reason="Log entry not found in history"
+            )
 
-        self.log(message, level="debug", jail=jail, indent=indent)
+        if log_entry.level != "screen":
+            raise libiocage.lib.errors.CannotRedrawLine(
+                reason=(
+                    "Log level 'screen' is required to redraw, "
+                    f"but got '{self.level}'"
+                )
+            )
 
-    def spam(self,
-             message,
-             jail=None,
-             indent=0):
+        line_number = self.PRINT_HISTORY.index(log_entry)
+        delta = len(self.PRINT_HISTORY) - line_number
 
-        self.log(message, level="spam", jail=jail, indent=indent)
+        # ToDo: Handle redrawig of multiline entries with different line count
+
+        output = "".join([
+            #"\033[s",            # SCP - Save Cursor Position
```

Comment:
If this is used (assuming you left a comment on it because it was useful), why is it commented out then?

Reply:
I was playing with various techniques to alter already printed output. Didn't get it to work as expected, but it was the most promising solution when it comes to multi-line editing